pyk/src/pyk/kast.py

```diff
@@ -1105,6 +1137,118 @@ def let_att(self, att: KAtt) -> 'KDefinition':
         return self.let(att=att)
 
 
+@dataclass(frozen=True)
+class Subst(Mapping[str, KInner]):
+    _subst: FrozenDict[str, KInner]
+
+    def __init__(self, subst: Mapping[str, KInner] = {}):
+        object.__setattr__(self, '_subst', FrozenDict(subst))
+
+    def __iter__(self) -> Iterator[str]:
+        return iter(self._subst)
+
+    def __len__(self) -> int:
+        return len(self._subst)
+
+    def __getitem__(self, key: str) -> KInner:
+        return self._subst[key]
+
+    def __mul__(self, other: 'Subst') -> 'Subst':
+        return self.compose(other)
+
+    def __call__(self, term: KInner) -> KInner:
+        return self.apply(term)
+
+    def minimize(self) -> 'Subst':
+        return Subst({k: v for k, v in self.items() if v != KVariable(k)})
+
```

Comment:
`minimize()` is maybe not a good name, because it implies that it's giving a different representation of the same substitution.

This is fundamentally changing the substitution, and returning a different substitution. So maybe it should be `remove_identities(...)` instead.

I really want to hammer home the point that _substitutions are not functions_, they have interpretations as functions. They are bindings for variables. The substitutions `Subst({'X': KVariable('X')})` and `Subst({})` are not the same substitution.

Reply:
The name comes from here:

https://github.com/runtimeverification/k/blob/fefe02c92cd16a9440275cc0cba9598e5337ba31/pyk/src/pyk/kastManip.py#L781

We can call it another name, but I'd prefer something shorter than `remove_identities`.